# Arise Application Challenge
---

Please answer the questions shown below by completing the code within this notebook and filling in your final answers in the respective block of the arise Challenge Answers form. 

---

### Step 1 (this step does not have an answer in the answer form)

Please import the Pandas package. You will need to install the package if you do not already have them (`hint - google anaconda install pandas`).

In [3]:
# Add your code for Question 1 in this code block
import pandas as pd

---

### Step 2 (this step does not have an answer in the answer form)

Use Pandas to read in the csv titled `application_data.csv` as a Panda's dataframe. Once you have read in the data as a dataframe use Panda's `head()` function to show an overview of the data.

In [4]:
# Add your code for Question 2 in this code block
mydf = pd.read_csv("application_data.csv")
mydf.head()


customerID        loanId appilcationDate  LoanNumber  LoanAmount  \
0  839247458  302953235622      2018-12-21           1      7500.0   
1  217540860  302960101337      2018-12-23           1      7500.0   
2  522238092  303016988846      2019-01-08           1      7500.0   
3  096754469  302999605504      2019-01-04           1      7500.0   
4  410354345  302951945878      2018-12-21           1      7500.0   

   InterestRate  TermDays repaymentDueDate repaymentPaidDate  
0          30.0        30       2019-01-20        2019-01-16  
1          30.0        30       2019-01-22        2018-12-29  
2          30.0        30       2019-02-07        2019-01-08  
3          30.0        30       2019-02-03        2019-01-08  
4          30.0        30       2019-01-20        2019-01-17

You should now be able to see that the dataset contains information relating to loans that have been disbursed. Each row of the data is specific to a particular loan and displays the customer's ID that took out the loan, a unique loan ID, the date when the client applied for the loan, the loan number of the specific loan (how many loans the client has now taken out), the amount, interest rate, term days, the date when the loan repayment was due and the date the loan was actually repaid.

---

### Question 1

Convert `appilcationDate`, `repaymentDueDate` and `repaymentPaidDate` in the dataframes to datetimes. 

In [25]:
# convert to dates
mydf["applicationDate"] = pd.to_datetime(mydf["appilcationDate"])
repaymentDueDate = pd.to_datetime(mydf.repaymentDueDate)
repaymentPaidDate = pd.to_datetime(mydf.repaymentPaidDate)


mydf.head()

customerID        loanId appilcationDate  LoanNumber  LoanAmount  \
0  839247458  302953235622      2018-12-21           1      7500.0   
1  217540860  302960101337      2018-12-23           1      7500.0   
2  522238092  303016988846      2019-01-08           1      7500.0   
3  096754469  302999605504      2019-01-04           1      7500.0   
4  410354345  302951945878      2018-12-21           1      7500.0   

   InterestRate  TermDays repaymentDueDate repaymentPaidDate dateDifference  \
0          30.0        30       2019-01-20        2019-01-16         4 days   
1          30.0        30       2019-01-22        2018-12-29        24 days   
2          30.0        30       2019-02-07        2019-01-08        30 days   
3          30.0        30       2019-02-03        2019-01-08        26 days   
4          30.0        30       2019-01-20        2019-01-17         3 days   

  applicationDate  
0      2018-12-21  
1      2018-12-23  
2      2019-01-08  
3      2019-01-04  
4      2018-12-21

Now that the dates have been converted, How many `unique`data types are contained within the application dataset?

In [6]:
# data types
mydf.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customerID         1000 non-null   object 
 1   loanId             1000 non-null   object 
 2   appilcationDate    1000 non-null   object 
 3   LoanNumber         1000 non-null   int64  
 4   LoanAmount         1000 non-null   float64
 5   InterestRate       1000 non-null   float64
 6   TermDays           1000 non-null   int64  
 7   repaymentDueDate   1000 non-null   object 
 8   repaymentPaidDate  1000 non-null   object 
dtypes: float64(2), int64(2), object(5)
memory usage: 70.4+ KB


####    There are three(3) unique data types

### Question 2

Please calculate the `average` loan amount, interest rate and termdays for all the loans in the dataset? `Round` all number to two decimal places.

In [7]:
# 2a avg loan amount
loan_sum = 0
loan_count = 0
for i in mydf.LoanAmount:
    loan_count += 1
    loan_sum += i    

Average_loanAmount = "{0:.2f}".format(loan_sum/loan_count)
print("The average loan amount is : {}".format(Average_loanAmount))

The average loan amount is : 13828.08


In [8]:
# 2b avg interest rate
InterestRate_sum = 0
InterestRate_count = 0
for _ in mydf.InterestRate:
    InterestRate_count += 1
    InterestRate_sum += _
    
Average_InterestRate = "{0:.2f}".format(InterestRate_sum/InterestRate_count)
print("The average Interest Rate is : {}".format(Average_InterestRate))

The average Interest Rate is : 23.06


In [9]:
# 2c avg term days 
TermDays_count = 0
TermDays_sum = 0
for e in mydf.TermDays:
    TermDays_count += 1
    TermDays_sum += e
    
Average_TermDays = "{0:.2f}".format(TermDays_sum/TermDays_count)
print("The average Term Days is : {}".format(Average_TermDays))

The average Term Days is : 42.08


---

### Question 3

Please create a new column called `dateDifference` which needs to contains the number of days between when the loan was repaid and when it was due to be repaid. 

Count the number of loans that were paid more than 3 days late. This can be done by first creating a dataframe only containing the loans that were paid more than 3 days late. 

In [10]:
# date difference
mydf["dateDifference"] = repaymentDueDate - repaymentPaidDate

In [11]:
# number of loans more than 3 days late
count = 0
for i in mydf.dateDifference.dt.days:
    if i < -3:
        count += 1
print("{} loans were paid 3 days late".format(count))

145 loans were paid 3 days late


---

### Question 4

Please group the dataset by `TermDays` and calculate the `mean` loan amount and the `standard deviation` of loan amount within each term.

What is the mean loan amount in 30-day loans and what is the standard deviation of  loan amount in 90-day loans? Please give your answers to the nearest two decimal places. 

In [15]:
# mean loan amount in 30-day loans and standard deviation of loan amount in 90-day loans
grouped_df = mydf.groupby("TermDays")
for key, item in grouped_df:
    if key == 30:
        print(item.describe())
        
for key, item in grouped_df:
    if key == 90:
        print(item.describe())
   

       LoanNumber    LoanAmount  InterestRate  TermDays  \
count  560.000000    560.000000    560.000000     560.0   
mean     2.137500   7941.964286     28.562143      30.0   
std      2.059236   1606.039495      4.034005       0.0   
min      1.000000   3000.000000     12.500000      30.0   
25%      1.000000   7500.000000     30.000000      30.0   
50%      2.000000   7500.000000     30.000000      30.0   
75%      2.000000   7500.000000     30.000000      30.0   
max     23.000000  20000.000000     30.000000      30.0   

                dateDifference  
count                      560  
mean         -2 days +02:24:00  
std    28 days 23:34:12.305911  
min        -252 days +00:00:00  
25%          -2 days +00:00:00  
50%            2 days 00:00:00  
75%           11 days 00:00:00  
max           30 days 00:00:00  
       LoanNumber     LoanAmount  InterestRate  TermDays  \
count   55.000000      55.000000     55.000000      55.0   
mean     4.709091   37672.727273     10.645455     

#### The mean for the loan amount for 30-days loan is 7941.96
####  The standard deviation for the loan amount for 90-days loan is 18050.14


--- 

### Question 5

`Late` loans are defined as loans that were paid `more than 3 days` after their due date. Please find the average number of days that these loans were paid beyond their due date for each `week of the year`. This will require you to break all `appilcationDate's` down into the week of the year they were made and then perform a group by. 

What is the average number of days that `late` loans applied for in the second week of the year were paid late? Please sumbit your answer rounded to two decimal places. 

In [48]:
mydf["dateDifference_as_int"] = pd.Series(mydf.dateDifference.dt.days)

In [50]:
mydf["applicationDate"].dt.week
grouped_weeks = mydf.groupby(mydf["applicationDate"].dt.week)

for key, item in grouped_weeks:
    if key == 2 and item.dateDifference_as_int < -3:
        print(item.dateDifference)

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [13]:
# week of year and average number of late days


---

### The End

Thank you for completing the challenge. Please make sure you have submitted all of your answers to the arise Challenge Answers form. 